### PR TITLE
Initiate transactions export in server via download button

### LINKED ISSUE #1484: Extend transactions CSV export
Parent issue #689

With #1479 we provided a basic CSV export, you can download the visible rows. But we need to discuss how to approach a more advanced solution. Like full transactions export using a new server request or email report link like wc-admin.

This also depends on issue #1483 to decide if we create a custom implementation or opt for migrating to wc-admin report table.

We can check now the event `wcadmin_wcpay_transactions_download` to have more information and evaluate the priority of this issue.

### PR DESCRIPTION
**Pull request by naman03malhotra**
Fixes #1484 

#### Changes proposed in this Pull Request

Trigger request to server when requesting a download of transaction table while using date or type filters. Otherwise, only download transactions presented in the table view.

#### Testing instructions

Note this PR relies on this server PR: 1178-gh-Automattic/woocommerce-payments-server

- Via WP-Admin go to **Payments > Transactions**
- Click the download link: https://cloudup.com/ccse5g7T_wc
- Ensure a CSV containing the transactions listed in the table view immediately downloads
- Select any filter (i.e. before December 1)
- Ensure a notice appears letting you know an export will be emailed to you
- Confirm you receive an email containing a download link for an export containing transactions that meet the criteria
- Repeat steps 3-6 with different combinations of filters

-------------------

- [ ] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

**Comment by chickenn00dle** [comment]
Hey @naman03malhotra 👋 

I was asked to work on Automattic/woocommerce-payments#1484 this sprint. I see you've started work on this already and I wanted to see if I could continue from where you left off in this PR?

**Comment by naman03malhotra** [comment]
Hey @chickenn00dle, please feel free to take it forward. Please let me know if I can help with anything. 🙂

**Comment by chickenn00dle** [comment]
Noting that this is almost ready for review. I'm having trouble getting tests to pass though. I need to mock `useDispatch` [here](https://github.com/Automattic/woocommerce-payments/pull/2939/files#diff-f2e931a8af39829d8b9fc3498e917d81b1d4d97526f17b61ecccc701d22a5f47R37-R40), but it looks like this is causing some import issues that I can't work out. 

@naman03malhotra I'd love some help if you have any idea as to what might be going on 🙇

File: client/transactions/list/test/index.tsx
Comment (chickenn00dle) [inline client/transactions/list/test/index.tsx:32]: Just noting that I spent some time today trying to get to the bottom of this, but was unable to find a way to get this mocked without the workaround mentioned here: p1631120638022300/1631117771.018600-slack-C01RTUVR81K

If you know a better way to approach this, please let me know!

File: client/transactions/list/test/index.tsx
Comment (chickenn00dle) [inline client/transactions/list/test/index.tsx:420]: Similar to my comment above, this was causing issues with the `@wordpress/data` mock, and it seemed as though it might not be needed for the CSV tests, so I removed them altogether. Please let me know if they actually are needed for some reason and perhaps we can explicitly reset/clear the relevant mocks rather than all.

**Comment by chickenn00dle** [comment]
@naman03malhotra I'm not able to add you as a reviewer for this PR since you are the original author, so pinging you here to let you know this is ready for review. 🙇

**Comment by naman03malhotra** [comment]
Hey @chickenn00dle sorry for the delay in response, thank you for picking this up. I'll have a look at it today or tomorrow after our NIRL meetup.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (brianyu28) [inline includes/wc-payment-api/class-wc-payments-api-client.php:649]: Are these (`page`, `pagesize`, etc.) needed for the server endpoint export? It looks like they were removed from the query parameters in https://github.com/Automattic/woocommerce-payments/pull/2939/commits/9c560a63aecf2fb3de64d6f5dcceb97b8c0a8948.

**Review by brianyu28 — COMMENTED** [review]
Looks good, thanks @chickenn00dle! The export works as expected, though applying the filters didn't seem to affect which transactions were exported (I got all of the transactions every time, even when applying filters). I think that might be due to a snake case vs. camel case difference in the query parameters, but it's possible I'm doing something wrong.

I should also note that my test site has fewer than 100 transactions, so I wasn't able to test this on a large transaction export (which I know was a main motivation for this feature). Might be useful for someone who has a test site with more transactions to give this a try too.

File: client/data/transactions/resolvers.js
Comment (chickenn00dle) [reply client/data/transactions/resolvers.js:61]: Great catch @brianyu28 💯 

I've fixed this in https://github.com/Automattic/woocommerce-payments/pull/2939/commits/3b05b8ed3923572926845374a2e79ede30750ab4

File: client/transactions/list/index.tsx
Comment (chickenn00dle) [reply client/transactions/list/index.tsx]: Ah. Completely missed this. Removed in https://github.com/Automattic/woocommerce-payments/pull/2939/commits/3b05b8ed3923572926845374a2e79ede30750ab4

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (chickenn00dle) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: You're right. I removed these from the endpoint, and missed these when cleaning them up from this PR.

I've removed these in https://github.com/Automattic/woocommerce-payments/pull/2939/commits/3b05b8ed3923572926845374a2e79ede30750ab4

**Comment by chickenn00dle** [comment]
Thanks for the review @brianyu28! 

> I should also note that my test site has fewer than 100 transactions, so I wasn't able to test this on a large transaction export 

Agreed. I'm actually in the same boat. My test account also has < 100 transactions. Hopefully @naman03malhotra has a larger pool of transactions to work with 🤞

**Comment by dechov** [comment]
Feel free to link a test account of mine [acct_1I9iAWR5iLftU7Bj](https://dashboard.stripe.com/test/connect/accounts/acct_1I9iAWR5iLftU7Bj/activity) (on WCPay Dev) that happens to have well over 100.

File: client/transactions/list/test/index.tsx
Comment (naman03malhotra) [reply client/transactions/list/test/index.tsx:420]: >this was causing issues with the @wordpress/data mock

Were you facing the similar error as above when these lines were present?

**Review by naman03malhotra — COMMENTED** [review]
Thanks for taking this up @chickenn00dle, the code looks good to me. Though while testing, I am able to see the success notice that an email will be sent, but I didn't receive the actual email. I've switched to the server branch, am I missing a step? Please guide me on how to receive an email, I'd like to see the formatting and content.

**Comment by chickenn00dle** [comment]
Hey @naman03malhotra 👋 

Sorry for the delay, I was AFK last week 🙇 

Thanks for the review on this!

> I am able to see the success notice that an email will be sent, but I didn't receive the actual email.

If you are testing locally, then you can see the emails via the email logger on server: `http://localhost:8086/wp-admin/admin.php?page=wpml_plugin_log`: https://cloudup.com/c6A7VrJqw8S


Otherwise, if you are syncing the changes to sandbox, you should be receiving emails to whatever the redirect mail you set up [here](https://github.com/Automattic/woocommerce-payments-server/blob/f5ede6b18b665229e0a9084e4b26e46580ea76d5/dev/0-sandbox-sample.php#L35) is.

> since we are getting the file URL in response why don't we trigger an auto-download for them. This will save them a trip from visiting their emails. We can modify the success message accordingly.

I've since changed the response from server to be something along the lines of:

```
{
    "exported_transactions": 123
}
``` 

The reason we are doing this is depending on the number of transactions that need to be fetched, this operation can potentially take some time. And since WC Admin already exports via email for other reports, I thought keeping this consistent would avoid any potential issues with waiting for long exports.

File: client/transactions/list/test/index.tsx
Comment (chickenn00dle) [reply client/transactions/list/test/index.tsx:420]: Yup. Specifically the following error:

```
TypeError: Cannot read property 'createNotice' of undefined
```

I suspect resetting/restoring all mocks is undoing the `@wordpress/data` mock hack implemented above. But I'm not entirely sure tbh. Any thoughts on this?

**Review by naman03malhotra — COMMENTED** [review]
Looks good to me, great work, tested works well. Thanks for working on it :)
+1 from my side.
ps: can't approve it, since I created this PR.

**Comment by chickenn00dle** [comment]
Thanks @naman03malhotra! 

I'll wait for a final look from @brianyu28 (and the server PR) before merging then.

**Review by brianyu28 — CHANGES_REQUESTED** [review]
Looks good, @chickenn00dle! In doing some additional testing, I noticed that if I turn on multiple transaction filters, the CSV export only includes transactions that match all of the filters, even if I have the page set to **Transactions match [ Any ] filters**. I'm guessing this could be fixed by passing the [`match`](https://github.com/Automattic/woocommerce-payments/blob/develop/client/data/transactions/resolvers.js#L22) field of `getQuery()` to `getTransactionsCSV` if it exists?

Other than that, everything's working well and looks ready to me.

**Comment by chickenn00dle** [comment]
Nice catch @brianyu28! I've fixed this in https://github.com/Automattic/woocommerce-payments/pull/2939/commits/b5d24d1ac7ff7d58af4f2284a789a457a01bca64

**Review by brianyu28 — APPROVED** [review]
Works great now, @chickenn00dle! Looks ready to merge once the server PR merges.

**Comment by chickenn00dle** [comment]
Thanks for the review y'all 🙌 

The server PR has been merged so I'll go ahead and merge this one now.

**Comment by chickenn00dle** [comment]
Just noting that I am reverting this for now since there were issues deploying the server PR and I have to step away for the evening. I'll have a further look into this tonight or over the weekend.

**Pull request merged**
